atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java

```diff
@@ -237,81 +261,99 @@ public SerializableTransactionManager serializable() {
                 userAgent());
 
         KvsProfilingLogger.setSlowLogThresholdMillis(config.getKvsSlowLogThresholdMillis());
-        KeyValueService kvs = ProfilingKeyValueService.create(rawKvs);
-        kvs = SweepStatsKeyValueService.create(kvs,
-                new TimelockTimestampServiceAdapter(lockAndTimestampServices.timelock()));
-        kvs = TracingKeyValueService.create(kvs);
-        kvs = AtlasDbMetrics.instrument(KeyValueService.class, kvs, MetricRegistry.name(KeyValueService.class));
-        kvs = ValidatingQueryRewritingKeyValueService.create(kvs);
-
-        SchemaMetadataService schemaMetadataService = SchemaMetadataServiceImpl.create(rawKvs,
+        KeyValueService keyValueService = initializeCloseable(() -> {
+            KeyValueService kvs = atlasFactory.getKeyValueService();
+            kvs = ProfilingKeyValueService.create(kvs);
+            kvs = SweepStatsKeyValueService.create(kvs,
+                    new TimelockTimestampServiceAdapter(lockAndTimestampServices.timelock()));
+            kvs = TracingKeyValueService.create(kvs);
+            kvs = AtlasDbMetrics.instrument(KeyValueService.class, kvs, MetricRegistry.name(KeyValueService.class));
+            return ValidatingQueryRewritingKeyValueService.create(kvs);
```

Comment:
Note that if one of the KVS objects creation fail midway through, we won't clean the initially created objects. We could work around this, but currently none of the created KVSs can fail after the first one, so... it should be fine.

Reply:
You can always fail with `OutOfMemoryError` or a messed-up classpath leading to linkage errors... 😛

I think this covers the majority of cases though, we can protect against these weird cases if this resurfaces in one of these directions in prod